compiler/src/middle_end/linearize.re

```diff
@@ -1375,46 +1350,22 @@ let rec transl_anf_statement =
         None;
       };
     let (exp_ans, exp_setup) = transl_comp_expression(~name?, vb_expr);
+    let tmp = gensym("destructure_target");
+    let destructure_setup = [BLet(tmp, exp_ans, Nonglobal)];
     let (rest_setup, rest_imp) =
       transl_anf_statement({
         ...s,
         ttop_desc: TTopLet(export_flag, Nonrecursive, mut_flag, rest),
       });
     let rest_setup = Option.value(~default=[], rest_setup);
     let setup =
-      switch (vb_pat.pat_desc) {
-      | TPatVar(bind, _)
-      | TPatAlias({pat_desc: TPatAny}, bind, _) =>
-        switch (mut_flag) {
-        | Mutable => [BLetMut(bind, exp_ans, Global({exported: exported}))]
-        | Immutable => [BLet(bind, exp_ans, Global({exported: exported}))]
-        }
-      | TPatTuple(_)
-      | TPatRecord(_) =>
-        let (tmp, anf_patts) =
-          bind_patts(
-            ~mut_flag,
-            ~global=Global({exported: exported}),
-            vb_pat,
-          );
-        [BLet(tmp, exp_ans, Global({exported: exported})), ...anf_patts];
-      | TPatAlias(patt, bind, _) =>
-        let binds =
-          switch (mut_flag) {
-          | Mutable => [
-              BLetMut(bind, exp_ans, Global({exported: exported})),
-            ]
-          | Immutable => [BLet(bind, exp_ans, Global({exported: exported}))]
-          };
-        let (tmp, anf_patts) =
-          bind_patts(~mut_flag, ~global=Global({exported: exported}), patt);
-        binds @ [BLet(tmp, exp_ans, Nonglobal), ...anf_patts];
-      | TPatAny => [BSeq(exp_ans)]
-      | _ =>
-        failwith(
-          "NYI: transl_anf_statement: Non-record or non-tuple destructuring in let",
```

Comment:
Do we want tests for this now that it seems that records and tuples can be destructured?

Reply:
Records and tuples could be destructured before, but this case was handling destructuing that wasn't one of those things (or patterns, constructors, etc.). I'll add some more tests :)